### PR TITLE
Make use of StorageHasher and HasherKind

### PR DESCRIPTION
**Pull request by kaichaosun**
Fix https://github.com/paritytech/substrate/issues/2449

### CODE REVIEW

**Comment by xlc** [comment]
Need new metadata version

**Comment by kaichaosun** [comment]
@xlc The signature to use double_map is not changed, seems no need to bump up the version of metadata.

**Comment by xlc** [comment]
The signature is changed. The fact you have to change the tests means the binary encoded value changed. Previously it is encoded as string but now encoded as an enum (a byte).

**Comment by bkchr** [comment]
@xlc is right and we need a new version.

CC @jacogr @amaurymartiny

**Comment by kaichaosun** [comment]
Thanks for the explanation. It makes sense to me.

File: srml/support/procedural/src/storage/impls.rs [hunk 1]
```diff
@@ -636,8 +636,11 @@ impl<'a, I: Iterator<Item=syn::Meta>> Impls<'a, I> {
 				}
 
 				fn key_for(k1: &#k1ty, k2: &#k2ty) -> Vec<u8> {
+					use #scrate::storage::hashed::generator::StorageHasher;
 					let mut key = #as_double_map::prefix_for(k1);
-					key.extend(&#scrate::Hashable::#k2_hasher(k2));
+					let mut key2 = vec![];
+					#scrate::codec::Encode::encode_to(k2, &mut key2);
+					key.extend(#scrate::#k2_hasher::hash(&key2[..]).to_vec());
```
Comment (bkchr) [inline srml/support/procedural/src/storage/impls.rs]: Pleas use something like
```
#scrate::codec::Encode::using(k2, |e| key.extend(&#scrate::#k2_hasher::hash(e))
```

Important bits are `using_encoded` and drop `to_vec()`.

File: srml/metadata/Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "srml-metadata"
-version = "2.0.0"
+version = "2.0.1"
```
Comment (bkchr) [inline srml/metadata/Cargo.toml]: No this is not what we requested, you need to alter the structures.

https://github.com/paritytech/substrate/blob/5088c0067e6c21f62c7412ab1d995497121cf595/srml/metadata/src/lib.rs#L315 This needs to be changed to `V5` and `V4` needs to be changed as for the other versions.

**Comment by kaichaosun** [comment]
I have finished the changes, please help review it. @xlc @bkchr @thiolliere

**Comment by amaurymartiny** [comment]
The JS part is being worked on cc @Stefie @axelchalon

**Pull request merged**